### PR TITLE
Strengthen partially supported invariant of base for chained sources

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103386
* #103456
* #103450
* __->__ #103445

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim

### CODE REVIEW

File: torch/_dynamo/source.py [hunk 1]
```diff
@@ -376,32 +368,30 @@ def name(self):
 
 
 @dataclasses.dataclass(frozen=True)
-class SuperSource(Source):
+class SuperSource(ChainedSource):
     type: Source
-    obj: Source
```
Comment (ezyang) [inline torch/_dynamo/source.py:381]: This BC breaking change feels like a bug farm. I suggest adding `kw_only`

File: torch/_guards.py [hunk 2]
```diff
@@ -558,6 +559,12 @@ def is_nn_module(self) -> bool:
         return self.guard_source().is_nn_module()
 
 
+# Subclasses can be found in torch/_dynamo/source.py
```
Comment (ezyang) [inline torch/_guards.py:562]: Explain a bit more, so people adding new sources know whether or not to inherit from this?

One question especially, is if I have a Source that contains multiple sources (as is the case with SuperSource), which one applies? I think this is why I wasn't a big fan of your proposed invariant here the last time we talked about it, to me Source is an AST (tree), there's no reason for it to have linked list structure.

Comment (voznesenskym) [reply torch/_guards.py:562]: Sources are trees.

Every source with a base should have just a single base, with the exception of SuperSource.

SuperSource is the only one - and it does not feel worth it to do `base -> [Source]` just for that - I'd rather carve out an exception for it.

File: torch/_dynamo/source.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/source.py:381]: fwiw this class is barely used - we could always make another subclass of MultiBaseSource lol

**Review by ezyang — APPROVED** [review]

File: torch/_guards.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_guards.py:562]: Added a comment.

File: torch/_dynamo/source.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/source.py:381]: Added a comment.

**Pull request closed without merging**